File: Xin/LeetCode/data_structure/chapterOne/sectionThree.py
"""
剑指 Offer 06. 从尾到头打印链表
输入一个链表的头节点，从尾到头反过来返回每个节点的值（用数组返回）。

示例 1：

输入：head = [1,3,2]
输出：[2,3,1]

限制：

0 <= 链表长度 <= 10000

作者：Krahets
链接：https://leetcode-cn.com/leetbook/read/illustration-of-algorithm/5dt66m/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
"""

# 递归法
class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        return self.reversePrint(head.next) + [head.val] if head else []

# 辅助栈法
class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        stack = []
        while head:
            stack.append(head.val)
            head = head.next
        return stack[::-1]

File: Xin/C++/LeetCode/data_structure/chapterOne/sectionThree.cpp
/*
 * @Author: your name
 * @Date: 2021-02-18 09:18:52
 * @LastEditTime: 2021-02-18 09:25:09
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \C++\LeetCode\data_structure\chapterOne\sectionThree.cpp
 */
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct ListNode
{
    int val;        // 节点值
    ListNode *next; // 后继节点引用
    ListNode(int x) : val(x), next(NULL) {}
};

//递归法
class Solution
{
public:
    vector<int> reversePrint(ListNode *head)
    {
        recur(head);
        return res;
    }

private:
    vector<int> res;
    void recur(ListNode *head)
    {
        if (head == nullptr)
            return;
        recur(head->next);
        res.push_back(head->val);
    }
};

//辅助栈法
class Solution
{
public:
    vector<int> reversePrint(ListNode *head)
    {
        stack<int> stk;
        while (head != nullptr)
        {
            stk.push(head->val);
            head = head->next;
        }
        vector<int> res;
        while (!stk.empty())
        {
            res.push_back(stk.top());
            stk.pop();
        }
        return res;
    }
};